### PR TITLE
better eclipse/m2e compatibility

### PR DESCRIPTION
**Pull request by jesboat**
(1/3) eclipse/m2e compatibility: trivial changes

1. RecordingHiveMetastore.java had two places where it's doing

    CacheBuilder.<K, V>newBuilder().build()

even though `newBuilder()` isn't generic. I was surprised to find that
JLS explicitly allows this (JLS 11th ed, sec 15.12.2.1), but ecj issues a
warning. Move the `<K, V>` to `build()` which actually is generic.

2. In TestArrayOperators and TestMapOperators, there are two places
where ecj is doing the inference differently from javac. I *suspect* ecj
is wrong here because, while I'm not curious enough to dredge through
JLS chapter 18, the fact that it reproduces in
https://gist.github.com/jesboat/90bda38a0f2d2208a8b114f80943ad34
only if `somehowRequired()` is present means at least something wonky is
going on.

---

(2/3) eclipse/m2e compatibility: remove test-jar dependency

Currently, pesto-tests scope compile has a dependency on presto-main's
test-jar. Unfortunately, eclipse/m2e can't handle this properly
(https://bugs.eclipse.org/bugs/show_bug.cgi?id=537851). All the
workarounds suck.

There are already a handful of testing utility classes in
presto-main/src/main/java/com/facebook/presto/testing/; the last
objectionable workaround (IMO) would be to move
com.facebook.presto.tests.QueryTemplate and
com.facebook.presto.util.StructuralTestUtil (not to be confused with
with com.facebook.presto.tests.StructuralTestUtil, grr) from
presto-main/src/test to presto-main/src/main and, for consistency, put
them in the com.facebook.presto.testing package along with the other
classes there.

This commit does the first part: move the files from src/test to
src/main and change the test-jar dependency to test scope (which m2e
does support). Moving the classes into the com.facebook.presto.testing
package is split into the next commit so it's easier to understand the
changes.

---

(3/3) eclipse/m2e compatibility: move some classes

Move com.facebook.presto.tests.QueryTemplate and
com.facebook.presto.util.StructuralTestUtil into the
com.facebook.presto.testing package, for consistency with the other
test utility classes in presto-main/src/main/. See previous commit for
reasons.

### CODE REVIEW

**Comment by mbasmanova** [comment]
CC: @raghavsethi @nezihyigitbasi

**Comment by arhimondr** [comment]
@jesboat I suspect there's no chance i can convince you to use Intellij? (You can enable eclipse shortcuts there! yay!)

ECJ is buggy. And i'm not sure how well does it support Java 11, that we may like to move to in the near future. From one side the changes you are trying to do don't hurt. From the other side eventually we may still decide that ECJ is not something we would like to support.

So, i kinda don't have a strong opinion here.

Do we want to support ECJ and the Eclipse IDE in general?

@mbasmanova / @rschlussel / @raghavsethi / @nezihyigitbasi  wanna hear your thoughts?

**Comment by rschlussel** [comment]
I agree.  You're better off using intellij and I don't think it's worth doing anything questionable or drastic to support eclipse if that were to become necessary later on (i.e. I wouldn't count on continued support).  However, since these changes seem fine to me, it makes sense to me to make Presto less tied to a particular tool.

**Comment by jesboat** [comment]
Welll.... I'll probably try to keep with Eclipse unless forces me off :)

I agree that it's not worth doing anything drastic to support eclipse/m2e/ecj. If, for example, ecj doesn't support Java 11 (I think it does), then I would say Presto should go to 11 and leave ecj behind. Or if Presto wanted to start including Kotlin (where IntelliJ is probably way ahead of Eclipse), I'd probably jump ship.

Since these are very localized changes (16 lines and two `git mv`s; the entire rest of the PR is renaming the two classes for stylistic consistency with the other classes in presto-main/src/main), I'm inclined to land this. (Will wait until Monday though.)

I have no expectation that anyone should do anything special to support ecj/eclipse/m2e (unless a bunch more people start using it for some reason, which seems unlikely.)

> ECJ is buggy. 

so is javac :P (e.g. https://github.com/facebook/fbthrift/blob/2785bea1334076eb9a2a2fc006beae0d9fcfc6cf/thrift/lib/java/src/main/java/com/facebook/thrift/UserExceptionHandler.java#L52)

**Comment by arhimondr** [comment]
@jesboat If you feel strong about using eclipse than it is fine. We can have these changes in.  

Just make sure that

- Commit messages are formatted by our code style and commit style: https://chris.beams.io/posts/git-commit/
- Second and third commits are squashed together. Use more commits for better readability is great, but usually before the merge we squash the commits, so we know that every commit compiles (and ideally passes the test)
- Travis tests are passing

And BTW, are those the only changes needed to support Eclipse?

**Comment by jesboat** [comment]
@raghavsethi, could you clarify why you requested changes? The only things I could see are minor violations of the subject line rules in https://chris.beams.io/posts/git-commit/, which seem relatively insignificant compared to the increased clarity of the subjects (specifically: in a shortlog, the commits clearly indicate that they're part of a series with a specific goal, where they are in the series, and what part of the goal they achieve.)

> we will likely freeze the repo at 5pm today until further notice (for the release) so you will not be able to land on Monday either.

Is there somewhere that describes this process more? (I'm trying to understand why we aren't just using branches for this, or if we could.)

**Comment by jesboat** [comment]
@arhimondr,

> Second and third commits are squashed together. Use more commits for better readability is great, but usually before the merge we squash the commits, so we know that every commit compiles (and ideally passes the test)

All of them compile individually. I suspect all of them would pass tests individually too, though I'm not sure if Travis actually tested this (or how to make Travis test them individually.)

> And BTW, are those the only changes needed to support Eclipse?

Yes.

**Comment by raghavsethi** [comment]
@jesboat: I would recommend you talk to @arhimondr or @rschlussel in person about our process.

**Comment by arhimondr** [comment]
@jesboat 

If i understand correctly, you are moving the files in `2/3` and than changing the package names in `3/3`. If that's the case i doubt the `2/3` commit would compile. Please correct me if I'm wrong, and got something incorrectly.

Also the commit `3/3` has merge conflicts committed in, I don't think the commit `3/3` will compile either.

> The only things I could see are minor violations of the subject line rules in

We are doing our best to follow that guide for all our commit messages. It makes it way easier to read the commit logs.

> which seem relatively insignificant compared to the increased clarity of the subjects

The describtions are good. The only thing is to hard-wrap them with 72 symbols limit.


About the titles

> (1/3) eclipse/m2e compatibility: trivial changes

I would rephrase it to "Make the codebase compatible with the EJC".

This commit is pretty independent from the following two. And, for example, can be reverted separately from the remaining two commits. I don't see the reason why should we link these commit with the following two. 

> (2/3) eclipse/m2e compatibility: remove test-jar dependency 
> (3/3) eclipse/m2e compatibility: move some classes 

This actually seems to be good to put as a single commit. As it does a single thing. And those two commits in separate wouldn't compile. I would squash those two guys, and call it something like: "Make the codebase compatible with the Eclipse IDE"
